viam-cartographer_internal_test.go

```diff
@@ -0,0 +1,348 @@
+	clientBuffer := make([]byte, m.clientBufferSize)
+	n, err := m.reader.Read(clientBuffer)
+	if err != nil {
+		return nil, err
+	}
+
+	resp := &v1.GetPointCloudMapResponse{
+		PointCloudPcdChunk: clientBuffer[:n],
+	}
+	return resp, nil
+}
+
+type internalStateClientMock struct {
+	grpc.ClientStream
+	reader           *bytes.Reader
+	clientBufferSize int
+}
+
+func makeInternalStateClientMock(b *[]byte, clientBufSize *int) *internalStateClientMock {
+	return &internalStateClientMock{reader: bytes.NewReader(*b), clientBufferSize: *clientBufSize}
+}
+
+func (m *internalStateClientMock) Recv() (*v1.GetInternalStateResponse, error) {
+	clientBuffer := make([]byte, m.clientBufferSize)
+	n, err := m.reader.Read(clientBuffer)
+	if err != nil {
+		return nil, err
+	}
+
+	resp := &v1.GetInternalStateResponse{
+		InternalStateChunk: clientBuffer[:n],
+	}
+	return resp, nil
+}
+
+func makeGetPositionResponse(
+	pose *commonv1.Pose,
+	quat *map[string]interface{},
+	componentReference *string,
+	quatKey string,
+) (*v1.GetPositionResponse, error) {
+	extra, err := structpb.NewStruct(map[string]interface{}{quatKey: *quat})
+	if err != nil {
+		return nil, err
+	}
+
+	resp := &v1.GetPositionResponse{
+		Pose:               pose,
+		ComponentReference: *componentReference,
+		Extra:              extra,
+	}
+	return resp, nil
+}
+
+func makeQuaternionFromProto(quat map[string]interface{}) spatialmath.Orientation {
+	return &spatialmath.Quaternion{
+		Real: quat["real"].(float64),
+		Imag: quat["imag"].(float64),
+		Jmag: quat["jmag"].(float64),
+		Kmag: quat["kmag"].(float64),
+	}
+}
+
+func TestGetPositionEndpoint(t *testing.T) {
+	svc := &cartographerService{}
+	mockSLAMClient := &inject.SLAMServiceClient{}
+	svc.clientAlgo = mockSLAMClient
+
+	// Add successful GetPosition client
+	mockSLAMClient.GetPositionFunc = func(
+		ctx context.Context,
+		in *v1.GetPositionRequest,
+		opts ...grpc.CallOption,
+	) (*v1.GetPositionResponse, error) {
+		resp, err := makeGetPositionResponse(&inputPose, &inputQuat, &inputComponentRef, "quat")
+		return resp, err
+	}
+
+	t.Run("origin pose success", func(t *testing.T) {
+		inputPose = commonv1.Pose{X: 0, Y: 0, Z: 0, OX: 0, OY: 0, OZ: 1, Theta: 0}
+		inputQuat = map[string]interface{}{"real": 1.0, "imag": 0.0, "jmag": 0.0, "kmag": 0.0}
+		inputComponentRef = "componentReference"
+		pose, componentRef, err := svc.GetPosition(context.Background())
+		test.That(t, err, test.ShouldBeNil)
+		expectedPose := spatialmath.NewPose(
+			r3.Vector{inputPose.X, inputPose.Y, inputPose.Z},
+			makeQuaternionFromProto(inputQuat),
+		)
+		test.That(t, pose, test.ShouldResemble, expectedPose)
+		test.That(t, componentRef, test.ShouldEqual, inputComponentRef)
+	})
+
+	t.Run("non origin pose success", func(t *testing.T) {
+		inputPose = commonv1.Pose{X: 5, Y: 5, Z: 5, OX: 1, OY: 0, OZ: 1, Theta: 0}
+		inputQuat = map[string]interface{}{"real": 1.0, "imag": 0.0, "jmag": 0.0, "kmag": 0.0}
```

Comment:
[q] Why did you change the orientation of the pose but not the quaternion? Why do we use a pose and a quaternion, when the pose already contains the orientation?

Reply:
The pose's orientation is ignored by our GetPosition function. This is because cartographer returns the orientation as a quaternion but our proto message defines it as an `OrientationVector`. Since the rust SDK spatial library is not yet available we are returning the orientation in the extra field to be parsed and used by the `spatialmath` lib in RDK.

Reply:
That being said I have updated the test to change the used quaternion and leave the unused oreintationvector in pose constant